### DISCUSSION TITLE
How to consume tRPC End API from separated App using trpcClient

### DISCUSSION THREAD

**Discussion by fortezhuo** [trpc/trpc#2496] [Q&A] [answered]
Hi,

Let say we have 2 separated apps (not in monorepo) eg : Employee Master Data App and Attendance App. Both apps developed using NextJS + tRPC + Prisma.

In Attendance App, we can't do import AppRouter from Employee Master Data App cause both apps are separated package
 
I know we have trpc-openapi for restful solution for consuming APIs. But can Attendance App consumes Employee Master Data e2e API's using trpcClient ?

**Comment by KATT** [comment, marked as answer]
You can publish the types of your backend as an npm package & import it in the client(s)

**Reply by fortezhuo** [reply]
Any idea how to extract types only as separated package ?

**Reply by fortezhuo** [reply]
I solve this problem by generate d.ts files separately

1. Create new tsconfig for build d.ts only
```
  "compilerOptions": {
    "declaration": true,
    "noEmit": false,
    "emitDeclarationOnly": true,
    "outDir": "./type"
  },
```
2. Publish the package, or copy the type folder if don't want to publish

3. VSCode Intellisense will be worked again 
![image](https://user-images.githubusercontent.com/9128724/186405379-dc045e00-ea36-4428-a633-8c85fb4910cf.png)

**Reply by joaquimnetocel** [reply]
It is worth noting that for this solution to work, the @trpc/server package must also be installed on the client. At least, that's how Eslint stopped complaining here.

Besides that, I think that it is not necessary to create another tsconfig.json. I just added 

"declaration": true,
"declarationDir": "./type",

to "compilerOptions" in "tsconfig".

**Reply by divramod** [reply]
@fortezhuo could you maybe share your package json configurations?
i am currently trying to also export my types as package.

**Reply by fortezhuo** [reply]
Here tsconfig.json for build the app

```
{
  "compilerOptions": {
    "strict": true,
    "resolveJsonModule": true,
    "module": "ESNext",
    "moduleResolution": "Node",
    "target": "ESNext",
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "types": ["vite/client"],
    "jsx": "react-jsx",
    "skipLibCheck": true,
    "esModuleInterop": true,
    "baseUrl": ".",
    "paths": {
      "@/*": ["src/*"]
    }
  },
  "exclude": [
    "dist",
    "node_modules",
    "./node_modules",
    "./node_modules/*",
    "./node_modules/@types/node/index.d.ts"
  ]
}
```
tsconfig.type.json for emit the type
```
{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "module": "ESNext",
    "moduleResolution": "Node",
    "declaration": true,
    "declarationDir": "./@db/btr",
    "skipLibCheck": true,
    "emitDeclarationOnly": true
  },
  "include": ["src/server/trpc/routers/index.ts", "@types"]
}
```
package.json script

```
    "build:type": "tsc --project ./tsconfig.type.json",
```

I generate different package.json only for storing the emit declaration.
```
{
  "name": "@db/btr",
  "type": "module",
  "version": "0.0.1",
  "types": "index.d.ts"
} 
```
Hope it helps solve your issues

**Comment by andrewalex** [comment]
@KATT Any chance of pinning this, or made easier to locate? Could also maybe a little note [here](https://trpc.io/docs/example-apps#vanilla-standalone-server-and-nodejs-procedure-calls) suggesting this method.  (Should this method be suggested, or is openapi better?)

Just thinking outloud, I wonder how wise it would be to set it up so the d.ts file is downloadable from the server letting it be an easy build step. (tho maybe some typescript magic would let you do this without the d.ts file)

**Reply by KATT** [reply]
Feel free to do a PR to the docs and/or the FAQ
